Various enhancements that remove the need to do 'vertical' flip - 100% support for rotation and display ordering.

lynniemagoo opened the pull request:
I started using this module for 7219 displays.  I had a single matrix unit purchased from Amazon which was mounted in such a way that when I put 2 units side-by-side, the algorithm for the module did not work as expected.   I have since purchased a second unit which is 4 matrices side-by-side.  What I found is that this module will properly display but I have to set 'flip' to 'vertical'.

While attempting to get the single matrix to work, I noticed that the 'options' supported a 'rotate' value.   I needed to use either 90 or 270 for this but quickly noticed that a value for 'degrees' was ignored in the default transformations algorithm.  When I tried adjusting the algorithm, I could not find a combination that would work for all types of matrices.

As a result of this, I have converted the 'flip' and 'rotate' functions into a single 'transform' function

`transform (matrix, degrees, direction)
`
The default implementation behaves exactly as it did before these changes.  If a user desires, they can provide a constructor option to override the 'transform' option such that user code is completely in control of the transform algorithm.   In my case, for the single matrix, I am able to use the default implementation for 'flip' but required a custom 'rotate' implementation.

In addition to this change, I also modified the scroll method with a do while loop and added the ability to prepend some blank text via a scrollIn scroll option 

`  m.scroll("Hello World!",{scrollIn:true, loop:true, speed:75});
`

My hopes are that this change will be accepted as part of 1.1.0 version and that a new version can be pushed to NPM.

Regards,
LynnieMagoo

lynniemagoo committed:
Added ability to provide customer transform algorithm to support displays that are rotated 90 degrees

lynniemagoo committed:
Updated Markdown with information about transform function and new 'scrollIn' scroll option.

lynniemagoo committed:
Rotation Enhancements including ability to reverse display order - provides additional capabilities when using rotate/flip to support mirrored scrolling displays

lynniemagoo committed:
Updating Examples based on new rotate/display ordering capabilities

lynniemagoo committed:
Updating documentation based on new rotate/display ordering capabilities

lynniemagoo committed:
Rotation Enhancements including ability to reverse display order - provides additional capabilities when using rotate/flip to support mirrored scrolling displays

lynniemagoo committed:
Minor updates to correctly reverse display ordering when requested to do so.

lynniemagoo committed:
Merge pull request #1 from lynniemagoo/RotationEnhancements

Rotation enhancements

antony commented:
Looks reasonable to me! I'll have a play with it tomorrow and merge :)

lynniemagoo commented:
Did you get a chance to try this out?

antony commented:
Hey, I haven't yet no. I'm absolutely buried with work, but I'll try to
check it as soon as I can.

Apologies for the delay!

On Sun, 13 Dec 2020 at 19:48, lynniemagoo <notifications@github.com> wrote:

> Did you get a chance to try this out?
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/antony/MAX7219.js/pull/1#issuecomment-744058724>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABVORNGW3JNHDX6KFWLEMTSUUK7RANCNFSM4UNFJR2Q>
> .
>
-- 
________________________________
ꜽ . antony jones . http://www.enzy.org

lynniemagoo committed:
Updates to make the character matrix read match that of the i2c-oled module (bit array needed to be reversed before parsing).  This allowed removal of the 7-i index when re-encoding the matrix.  The side-effect of this change is that a user will need to set flip to vertical, modify rotation, or change display ordering.  However, this is the cleanest approach to support consistent font processing of other possible OLED fonts.

lynniemagoo committed:
Update Crypto example to no longer require vertical flip due as bit direction is no longer reversed (see previous commit note).

lynniemagoo commented:
I made another minor update along with changes to the examples to illustate how to use the new functionality. One internal change was how the character arrays are read from the font. I had previously reversed bits in encodeByte to be consistent however, what I found is that the chartoMatrix function was actually reading bits in reverse order as compared to other displays based on OLED fonts. Therefore, I reverted a previous change in encodeByte in favor of a change to charToMatrix (added .reverse()). 

Upon doing this, I had to flip the 0 and 180 degree transformation rotation algorithms and adjust examples accordingly.

The end result is a very robust module that at some point, might be able to other OLED fonts that are <= 8x8.
     On Sunday, December 13, 2020, 01:59:25 PM CST, Antony Jones <notifications@github.com> wrote:  
 
 

Hey, I haven't yet no. I'm absolutely buried with work, but I'll try to
check it as soon as I can.

Apologies for the delay!

On Sun, 13 Dec 2020 at 19:48, lynniemagoo <notifications@github.com> wrote:

> Did you get a chance to try this out?
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/antony/MAX7219.js/pull/1#issuecomment-744058724>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABVORNGW3JNHDX6KFWLEMTSUUK7RANCNFSM4UNFJR2Q>
> .
>
-- 
________________________________
ꜽ . antony jones . http://www.enzy.org

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or unsubscribe.

lynniemagoo committed:
Add support for scrollOptions::scrollIncrement which defaults to 1.
Add new 'display' method which will display a string of up to controllerCount characters with optional start and length parameters.
Remove previous change reversing the bit ordering of bytes pulled from the array.  This was reversed based on reviewing the physical layout of the 4-matrix module I purchased from Amazon US.

lynniemagoo committed:
Add support for scrollOptions::scrollIncrement which defaults to 1.
Add new 'display' method which will display a string of up to controllerCount characters with optional start and length parameters.
Remove previous change reversing the bit ordering of bytes pulled from the array.  This was reversed based on reviewing the physical layout of the 4-matrix module I purchased from Amazon US.

lynniemagoo committed:
Merge branch 'ObjectifyModule' of github.com:lynniemagoo/MAX7219.js into ObjectifyModule

Minor bugfix after recommit fail.

lynniemagoo committed:
Need MSB-LSB bit ordering reversal that was previously removed as part of this branch to guarantee backward compatibility with antony's original code around the 'set' method.

lynniemagoo committed:
Update README to document the scrollIncrement field of scroll() options.

lynniemagoo committed:
Merge pull request #2 from lynniemagoo/ObjectifyModule

Objectify module.
Add scrollIncrement as scroll() option.
Add display() function to support display of short strings.
Move external functions inside MAX7219 class.

lynniemagoo commented:
I made some additional changes tonight.  You can actually check the pull request done into my master branch to see the changes.

lynniemagoo committed:
Optimize display function to use letter() as the output is character-based to a specific display.
Enhance scroll() to support additional options 'glyphRotate' and 'glyphFlip' which are used on individual character matrices when building the bit array for scroll.  This allows you to display text and scroll up/down vertically as a single column.

J
a
v
a
s
c
r
i
p
t
!

const Max7219 = require('max7219-display')
const controllerCount = 4;
const m = new Max7219({ font:"cp437", device: '/dev/spidev0.0', reverseDisplayOrdering:false, controllerCount: controllerCount, flip: 'none', rotate:0})

async function main() {
  await m.reset()

  //...

  await m.scroll("Javascript!",{scrollIn:true, loop:false, speed:100, scrollIncrement:1, glyphRotate:270, glyphFlip:'none'});
}
main();

lynniemagoo committed:
Bugfix - display() was not calculating length of string properly for display when length was -1.

antony commented:
Hi again,

Please accept my further apologies for not having time to review this properly yet.

From looking over the code itself, The changes you have made look excellent!

I will schedule some time this weekend to go over it. I wish there were a way to emulate the matrix so that we could write tests for this sort of thing!

Thank you for your ongoing hard work.

Thanks,
Antony

lynniemagoo commented:
Antony,

I made some additional changes tonight to the new display() function and to the scroll() function. My check-in comments around scroll() options specifically glyphRotate and glyphFlip are used to adjust text offset prior to loading into the scroll matrix to facilitate scrolling text that is rotated 90 degrees.

I've made every attempt to preserve your 'set' functionality to load an array that is generated with the HTML page you created. Please also retest that functionality to ensure I didn;t break something there.

Thanks,
Lyndel


     On Wednesday, January 6, 2021, 03:18:44 AM CST, Antony Jones <notifications@github.com> wrote:  
 
 


Hi again,

Please accept my further apologies for not having time to review this properly yet.

From looking over the code itself, The changes you have made look excellent!

I will schedule some time this weekend to go over it. I wish there were a way to emulate the matrix so that we could write tests for this sort of thing!

Thank you for your ongoing hard work.

Thanks,
Antony

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or unsubscribe.

antony commented on examples/hello-scroll.js:
```diff
@@ -3,13 +3,14 @@
 const controllerCount = 4
 
-const m = new Max7219({ device: '/dev/spidev0.0', controllerCount, flip: 'vertical' })
+const m = new Max7219({ device: '/dev/spidev0.0', controllerCount, flip: 'none' })
```
we can probably just remove the flip here, the default should be `none`.

lynniemagoo replied:
All examples updated to change to resetAll() and to remove default option flags.

antony commented on examples/text.js:
```diff
@@ -1,13 +1,14 @@
+  //await m.reset(1)
+  //await m.reset(2)
+  //await m.reset(3)
+  await m.reset();
```
no need to comment, this is git :) just delete the non-required lines! great!

lynniemagoo replied:
Examples updated.  Please review.

antony commented on lib/index.js:
```diff
@@ -101,81 +41,248 @@ const scrollOptions = {
    async displayTest (controller, enable) {
-     setController(this, controller)
-     await shiftOut(this, Registers.DisplayTest, enable ? on : off)
+     if (undefined === enable) {
```
you can just use `if (!enable)` here. In terms of best practices, the best way to keep this block clean is to have the positive condition first, and drop the else:
```js
if (enable) {
  this.setController(controller)
  await this.shiftOut(Registers.DisplayTest, enable ? on : off)
  return
}

for (controller = 0; controller < this.controllerCount; controller++) {
  this.setController(controller)
  await this.shiftOut(Registers.DisplayTest, controller ? on : off)
}
```

lynniemagoo replied:
Functions were created.  Please review latest comments.

antony commented on lib/index.js:
```diff
@@ -101,81 +41,248 @@ const scrollOptions = {
-     await shiftOut(this, Registers.Shutdown, on)
-     await this.displayTest(controller, false)
-     await this.clear(controller)
+     if (undefined === controller) {
```
same with this block.

i'm a big believer in specific methods that describe their function so I might actually suggest just making two functions here - resetAll, and resetSingle, and resetAll just calls resetSingle - this should probably be done for all the functions where you check the parameter passed. It's unecessary complexity:

```js
async function resetSingle (controller) {
   this.setController(controller)
   await this.shiftOut(Registers.DecodeMode, 0)
   await this.shiftOut(Registers.ScanLimit, 7)
   await this.shiftOut(Registers.Shutdown, on)
   await this.displayTest(ctrl, false)
   await this.clear(controller)
}

async function resetAll () (
  for (controller = 0; controller < this.controllerCount; controller++) {
    await resetSingle(controller)
  }
}
```

lynniemagoo replied:
Agree.  I will create 3 functions.  reset() which calls resetSingle() and resetAll() which loops and calls resetSingle().  This will guarantee backward compatibility.   Thoughts?

If we agree then I'll apply the same pattern for displayTest(), clear(), etc.

lynniemagoo replied:
Functions were created.  Please review latest comments.

antony commented on lib/index.js:
```diff
@@ -101,81 +41,248 @@ const scrollOptions = {
-     const matrix = emptyMatrix()
-     await this.set(controller, matrix)
+     const matrix = Max7219.emptyMatrix()
+     if (undefined === controller) {
```
same with this one. I think two methods is ultimately more readable.

lynniemagoo replied:
Functions were created.  Please review latest comments.

antony commented on lib/index.js:
```diff
@@ -101,81 +41,248 @@ const scrollOptions = {
+   async display (text, start = 0, length = -1) {
+     const controllerCount = this.controllerCount;
+     // skip past start characters if start > 0
+     text = (start > 0) ?  text.substring(start) : text;
```
it's generally a bad idea to mutate parameters passed in by a user. this results in you changing data that they own, without their knowledge, and can result in very strange behaviour.

I would suggest assigning `text` to our own variable first and then modifying that as required.

lynniemagoo replied:
In Javascript, strings are immutable meaning that when we change text, the change is local to the function.  However, from a stylistic perspective I agree and will make the change here and in scroll.

lynniemagoo replied:
Length is also a primitive type and any changes are local to the function but for stylistic reason, I'll make the change.

lynniemagoo replied:
Although I don't fully agree based on primitive types and immutability, both display() and scroll have been updated to use local variables instead of the parameters passed into the function.

antony commented on lib/index.js:
```diff
@@ -101,81 +41,248 @@ const scrollOptions = {
+     text = (start > 0) ?  text.substring(start) : text;
+     // Ensure we don't have more characters than controllers and ensure if text is too short we pad to controllerCount
+     // length with spaces.
+     length = Math.min(controllerCount, (length >= 0 && length <= text.length) ? length : text.length);
```
same with this one

lynniemagoo replied:
Although I don't fully agree based on primitive types and immutability, both display() and scroll have been updated to use local variables instead of the parameters passed into the function.

antony commented on lib/transformations.js:
```diff
@@ -1,38 +1,65 @@
 
-  function horizontal (matrix) {
-    return matrix.map(r => r.reverse() )
+function flip (matrix, direction) {
```
I never enjoyed this block of code anyway. I might suggest a hash:

```suggestion
function flip (matrix, direction) {

  const matrixFunctions = {
    none: m => m,
    vertical: m => vertical(m),
    horizontal: m => horizontal(m),
    both: m => vertical(horizontal(m))
  }
  
  const flipFunction = matrixFunctions[direction]
  
  if (flipFunction) {
    return flipFunction(matrix)
  }
  
  throw new Error('You can only flip by none, horizontal, vertical, or both')
}
```

lynniemagoo replied:
Please review entire transformation.js file.  It did come out cleaner.  

As a matter of choice, always opt for putting the hash outside of the function to prevent the hash entry functions from having to be compiled each time the function is called.

antony commented on lib/transformations.js:
```diff
@@ -1,38 +1,65 @@
+    }
+    // Original font matrices encoded with 90 degree rotation in clockwise direction.
+    let rotated;
+    switch(degrees) {
```
you could use the same function as I've done with flipping above here - bearing in mind that a hash is indexed by strings so you would need to convert the parameter to a string first: `rotationFunctions(`${degrees}`)`

lynniemagoo replied:
I agree that a hash is indexed by strings.  However, my experience is that when you store a field into a JavaScript object, it first converts the key to a string.  So, to do this, I think the conversion would be automatic.  I will test.

lynniemagoo replied:
I was correct on my above comment regarding hash key type of string.  If you pass in an integer, it will be converted to a string prior to lookup in an Object.

Please review entire transformation.js file. I tested both rotation and flip from the examples.  I also added 2 new examples illustrating the glyphRotate/glyphFlip options of scroll.  Hoping you have a display with which to test.

antony reviewed (changes requested):
Some amazing work here! I've really only fed back on stylistic changes. The original author's code was pretty complex and I had to undo a lot of the decisions - I think in order to keep the simplicity (reducing the indent, keeping functions small, removing repeated code etc) it would be advisable to make these changes - but otherwise, awesome work!

lynniemagoo commented:
Anthony,

Thanks for the feedback.

I will look into this in more detail tomorrow and agree with your comments. I like the idea of function hashes for the transformations. Some notes below on why I did some of the changes the way I did.

In lib/index.js:

> + this.setController(controller)
+ await this.shiftOut(Registers.Intensity, brightness)
+ }
+ } else {
+ if (0 > brightness > 15) {
+ throw new Error('Brightness must be between 0 and 15')
+ }
+ this.setController(controller)
+ await this.shiftOut(Registers.Intensity, brightness)
+ }
+ }
+ 
+ async display (text, start = 0, length = -1) {
+ const controllerCount = this.controllerCount;
+ // skip past start characters if start > 0
+ text = (start > 0) ? text.substring(start) : text;
it's generally a bad idea to mutate parameters passed in by a user. this results in you changing data that they own, without their knowledge, and can result in very strange behaviour.

I would suggest assigning text to our own variable first and then modifying that as required.

I generally agree and will make the change as in C/C++ or other languages besides Java/Javascript, strings are not immutable . If we change 'text', we are not changing the original caller's variable, we get a local copy..



With regard to this one...
In lib/index.js:

> const zeroArray = new Array(8).fill(0)
- await shiftOut(this, Registers.DecodeMode, encodeByte(zeroArray))
- await shiftOut(this, Registers.ScanLimit, 7)
- await shiftOut(this, Registers.Shutdown, on)
- await this.displayTest(controller, false)
- await this.clear(controller)
+ if (undefined === controller) {
same with this block.

i'm a big believer in specific methods that describe their function so I might actually suggest just making two functions here - resetAll, and resetSingle, and resetAll just calls resetSingle - this should probably be done for all the functions where you check the parameter passed. It's unecessary complexity:

async function resetSingle (controller) {
 this.setController(controller)
 await this.shiftOut(Registers.DecodeMode, 0)
 await this.shiftOut(Registers.ScanLimit, 7)
 await this.shiftOut(Registers.Shutdown, on)
 await this.displayTest(ctrl, false)
 await this.clear(controller)
}

async function resetAll () (
 for (controller = 0; controller < this.controllerCount; controller++) {
 await resetSingle(controller)
 }
}

I toyed with the idea of a reset() and resetAll() as you suggested and opted for the single function as if user passes in a controller (!== undefined), we will use it. Otherwise, assumption is that user wants to reset all controllers. This also preserved backward compatibility.

Maybe to preserve backward compatibility, we have 3 functions: reset() which calls resetSingle(controller) and resetAll() which in a loop calls resetSingle()? Thoughts?
     On Saturday, January 9, 2021, 05:44:58 PM CST, Antony Jones <notifications@github.com> wrote:  
 
 


@antony requested changes on this pull request.

Some amazing work here! I've really only fed back on stylistic changes. The original author's code was pretty complex and I had to undo a lot of the decisions - I think in order to keep the simplicity (reducing the indent, keeping functions small, removing repeated code etc) it would be advisable to make these changes - but otherwise, awesome work!

In examples/hello-scroll.js:
> @@ -3,13 +3,14 @@ const Max7219 = require('..') const controllerCount = 4 -const m = new Max7219({ device: '/dev/spidev0.0', controllerCount, flip: 'vertical' })+const m = new Max7219({ device: '/dev/spidev0.0', controllerCount, flip: 'none' })
we can probably just remove the flip here, the default should be none.

In examples/text.js:
>   async function init () {-  await m.reset(0)-  await m.reset(1)-  await m.reset(2)-  await m.reset(3)+  //await m.reset(0)+  //await m.reset(1)+  //await m.reset(2)+  //await m.reset(3)+  await m.reset();
no need to comment, this is git :) just delete the non-required lines! great!

In lib/index.js:
>     }     async displayTest (controller, enable) {-     setController(this, controller)-     await shiftOut(this, Registers.DisplayTest, enable ? on : off)+     if (undefined === enable) {
you can just use if (!enable) here. In terms of best practices, the best way to keep this block clean is to have the positive condition first, and drop the else:
if (enable) {  this.setController(controller)  await this.shiftOut(Registers.DisplayTest, enable ? on : off)  return}enable = controller;for (controller = 0; controller < this.controllerCount; controller++) {  this.setController(controller)  await this.shiftOut(Registers.DisplayTest, enable ? on : off)}
In lib/index.js:
>       const zeroArray = new Array(8).fill(0)-     await shiftOut(this, Registers.DecodeMode, encodeByte(zeroArray))-     await shiftOut(this, Registers.ScanLimit, 7)-     await shiftOut(this, Registers.Shutdown, on)-     await this.displayTest(controller, false)-     await this.clear(controller)+     if (undefined === controller) {
same with this block.

i'm a big believer in specific methods that describe their function so I might actually suggest just making two functions here - resetAll, and resetSingle, and resetAll just calls resetSingle - this should probably be done for all the functions where you check the parameter passed. It's unecessary complexity:
async function resetSingle (controller) {   this.setController(controller)   await this.shiftOut(Registers.DecodeMode, 0)   await this.shiftOut(Registers.ScanLimit, 7)   await this.shiftOut(Registers.Shutdown, on)   await this.displayTest(ctrl, false)   await this.clear(controller)}async function resetAll () (  for (controller = 0; controller < this.controllerCount; controller++) {    await resetSingle(controller)  }}
In lib/index.js:
>     async clear (controller) {-     setController(this, controller)-     const matrix = emptyMatrix()-     await this.set(controller, matrix)+     const matrix = Max7219.emptyMatrix()+     if (undefined === controller) {
same with this one. I think two methods is ultimately more readable.

In lib/index.js:
> +         this.setController(controller)+         await this.shiftOut(Registers.Intensity, brightness)+       }+     } else {+       if (0 > brightness > 15) {+         throw new Error('Brightness must be between 0 and 15')+       }+       this.setController(controller)+       await this.shiftOut(Registers.Intensity, brightness)+     }+   }+   +   async display (text, start = 0, length = -1) {+     const controllerCount = this.controllerCount;+     // skip past start characters if start > 0+     text = (start > 0) ?  text.substring(start) : text;
it's generally a bad idea to mutate parameters passed in by a user. this results in you changing data that they own, without their knowledge, and can result in very strange behaviour.

I would suggest assigning text to our own variable first and then modifying that as required.

In lib/index.js:
> +     } else {+       if (0 > brightness > 15) {+         throw new Error('Brightness must be between 0 and 15')+       }+       this.setController(controller)+       await this.shiftOut(Registers.Intensity, brightness)+     }+   }+   +   async display (text, start = 0, length = -1) {+     const controllerCount = this.controllerCount;+     // skip past start characters if start > 0+     text = (start > 0) ?  text.substring(start) : text;+     // Ensure we don't have more characters than controllers and ensure if text is too short we pad to controllerCount+     // length with spaces.+     length = Math.min(controllerCount, (length >= 0 && length <= text.length) ? length : text.length);
same with this one

In lib/transformations.js:
>  -  function horizontal (matrix) {-    return matrix.map(r => r.reverse() )+function flip (matrix, direction) {
I never enjoyed this block of code anyway. I might suggest a hash:
⬇️ Suggested change-function flip (matrix, direction) {+function flip (matrix, direction) {++  const matrixFunctions = {+    none: m => m,+    vertical: m => vertical(m),+    horizontal: m => horizontal(m),+    both: m => vertical(horizontal(m))+  }+  +  const flipFunction = matrixFunctions[direction]+  +  if (flipFunction) {+    return flipFunction(matrix)+  }+  +  throw new Error('You can only flip by none, horizontal, vertical, or both')+}
In lib/transformations.js:
> - const rotated = matrix.map(r => r.slice()).reverse()- - for (let i = 0; i < rotated.length; i++) {-   for (let j = 0; j < i; j++) {-     let temp = rotated[i][j]-     rotated[i][j] = rotated[j][i]-     rotated[j][i] = temp-   }- }+function rotate(matrix, degrees) {+    if (![0, 90, 180, 270].includes(degrees)) {+        throw new Error('You can only rotate by 0, 90, 180, or 270 degrees.')+    }+    // Original font matrices encoded with 90 degree rotation in clockwise direction.+    let rotated;+    switch(degrees) {
you could use the same function as I've done with flipping above here - bearing in mind that a hash is indexed by strings so you would need to convert the parameter to a string first: rotationFunctions(${degrees})

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or unsubscribe.

lynniemagoo committed:
Update transformation based on feedback from antony.   Please review.

lynniemagoo committed:
Updates based on code review feedback from antony.

1. Provide unique functions for single controller, all controllers, and the original function defaults as before to single controller.

reset() -> reset(), resetSingle(), resetAll()
clear() -> clear(), clearSingle(), clearAll()
setIntensity() -> setIntensity(), setIntensitySingle(), setIntensityAll()
displayTest() -> displayTest(), displayTestSingle(), displayTestAll()

2. Updated both display() and scroll() to create local copies of variables passed in as both functions mutate the values.

lynniemagoo committed:
Updated examples based on code review by antony.

1. Removed commented lines and options that were default.
2. Updated examples to use resetAll() method vs reset().
3. Provided 2 additional examples illustrating how to use glyphRotate and glyphFlip options in scroll().

lynniemagoo commented:
Antony,  

Apologies if my comments above are mixed/mashed.  This was my first pull request in Github.  Been using git for years with BitBucket at my day job but recently started using both github/gitlab here at home for these node modules.

I have a request for testing to validate that we/I don't need to change your html matrix generator.  
I'm hoping you have a display and a pi to test this.  Can you please confirm that if you use a 4 or 1 matrix display modules and load a matrix generated by your HTML page, that it displays as expected; meaning that the user is not required to specify a flip or rotate?  I tried to do this multiple times but as I had both types of display modules and one of the types required rotate/flip and one did not.  As you wrote the code, you know the original intent/outcome.

Regarding order of the pull request review, may I suggest that you review in this order to make your life easier.

1. Various examples plus the new examples.
2. Tranformations.js
3. index.js

Thanks,
Lynnie

lynniemagoo commented:
Did you get a chance to review my changes?

antony commented:
Heya - further apologies, I'm absolutely snowed under right now. I'll give this another review tomorrow :) It's looking good!

With regards to your comment:

>> I'm hoping you have a display and a pi to test this. Can you please confirm that if you use a 4 or 1 matrix display modules and load a matrix generated by your HTML page, that it displays as expected; meaning that the user is not required to specify a flip or rotate?

I have 4-conjoined 8x8 matrix modules. I remember I had to do a lot of flipping to get them to work :)

>> I tried to do this multiple times but as I had both types of display modules and one of the types required rotate/flip and one did not. As you wrote the code, you know the original intent/outcome.

I'll validate this when I test!

antony commented:
Hi Lynnie!

Again apologies for how long it has taken for me to review this. I've set it up and had a go, and it's great!

I noticed that, in theory, all of the examples are upside down on my board - but that's also true for the original examples. I don't know if I ever decided whether up was down and down was up or if I just rotated the displays by hand - I think I did!

What I was more concerned about was the whole thing being broken or the code being messy, and none of this is the case!

As for the matrix builder I wrote - that also doesn't really have a concept of upside down or not, it just uses `.reverse()`, so I can easily just adapt that when I come to re-visiting that builder. I meant to extract it as a github hosted page, so it's due a look anyway.

Either way this looks great, I love what you've done, so I'll get this merged!

Thanks for your hard work and neverending patience :)
